src/main/host/syscall_handler.c

```diff
@@ -488,6 +492,21 @@ SysCallReturn syscallhandler_make_syscall(SysCallHandler* sys,
             break;
     }
 
+    // If we would be blocked, but there's a signal pending, interrupt. We do this check
+    // *after* trying the syscall so that in the case that a syscall was blocked and becomes
+    // both unblocked and signalled before running again, the syscall is allowed to complete.
+    // This should make it a little harder to lose data, and helps comply with e.g. "If an I/O call
+    // on a slow device has already transferred some data by the time it is interrupted by a signal
+    // handler, then the call will return a success  status  (normally,  the  number of bytes
+    // transferred)." from signal(7).
+    if (scr.state == SYSCALL_BLOCK &&
+        thread_unblockedSignalPending(sys->thread, sys->shimShmemHostLock)) {
+        SysCallCondition* condition = scr.cond;
+        utility_assert(condition);
+        syscallcondition_unref(condition);
+        scr = (SysCallReturn){.state = SYSCALL_DONE, .retval = -EINTR};
+    }
```

Comment:
To clarify, if thread A is blocked on a syscall, and thread B sends a signal to thread A, shadow immediately interrupts thread A to process the signal? And after processing the signal, thread A's syscall will return EINTR and shadow will discard the blocking syscall condition?

Reply:
Not quite. I tried to clarify the comment. In this particular scenario:

* thread A is blocked on a syscall
* thread B sends a signal to A, which sets a pending bit on A and schedules it to run
* thread B continues running until it is blocked or exits
* thread A runs, first retrying the original syscall it was blocked on.
* If the syscall completes then its value is allowed to be returned. In the shim, the return code will be saved on the stack, the signal handler will run, and then the syscall's return value is returned. Effectively we're emulating that the signal arrived just after the syscall completed.
* If the syscall is still blocked, then instead of blocking the thread again we return EINTR. As in the other case, the signal handler runs, and then EINTR is returned for the original syscall.